### PR TITLE
[WEB3-41] TA integration and Admin endpoints

### PR DESCRIPTION
**Pull request by goncaloadolfo**
# Description

This PR integrates the connection with team allocator api, to confirm that the user is active and also to get unit information. Additionally, it includes the protection for admin users only in some endpoints, such as creating a new giveaway, and participants management.

Ticket: https://runtime-revolution.atlassian.net/browse/WEB3-41

## Type of change

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## How Has This Been Tested?

1. Login with your account and verify the user saved in the database (check the unit)
2. Try to create a new giveaway as a normal user and verify that you don't have the right access
3. Verify that it is no longer possible to manage users when the giveaway is not active

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by marciammart — CHANGES_REQUESTED** [review]

File: packages/node-app/src/tests/giveaway.controller.test.ts [hunk 1]
```diff
@@ -193,8 +200,8 @@ describe('POST /giveaways', () => {
   );
 
   it(
-    'should not create new giveaway while not authenticated',
-    notAuthenticated(async () => {
+    'should not create new giveaway while authenticated',
```
Comment (marciammart) [inline packages/node-app/src/tests/giveaway.controller.test.ts:203]: To make it clearer why it should not create the giveaway even though the user is authenticated:
```suggestion
    'should not create new giveaway if user is not admin',
```

File: packages/node-app/src/tests/location.controller.test.ts [hunk 2]
```diff
@@ -44,24 +44,24 @@ describe('POST /locations', () => {
   );
 
   it(
-    'should not create new location while not authenticated',
-    notAuthenticated(async () => {
+    'should not create new location while authenticated',
```
Comment (marciammart) [inline packages/node-app/src/tests/location.controller.test.ts:47]:
```suggestion
    'should not create new location if user is not admin',
```

File: packages/node-app/src/tests/location.controller.test.ts [hunk 3]
```diff
@@ -107,8 +107,8 @@ describe('DELETE /locations/:id', () => {
   );
 
   it(
-    'should not delete location while not authenticated',
-    notAuthenticated(async () => {
+    'should not delete location while authenticated',
```
Comment (marciammart) [inline packages/node-app/src/tests/location.controller.test.ts:110]:
```suggestion
    'should not delete location if user is not admin',
```

File: packages/node-app/src/utils/auth.utils.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      headers: {
+        Authorization: `${tokenType} ${accessToken}`,
+        Accept: 'application/json',
+      },
+    });
+
+    if (res.status === 200) {
+      const userInfo = res.data;
+      const taSession = await TAClient.signIn();
+      const runtimePeople = await TAClient.getPeople(taSession, false);
```
Comment (marciammart) [inline packages/node-app/src/utils/auth.utils.ts:22]: Isn't there an endpoint to get only the information of the current user?

File: packages/node-app/src/utils/validations.utils.ts [hunk 5]
```diff
@@ -54,9 +64,12 @@ const validateUnit = (participant, giveaway) => {
   const requiredUnit = giveaway.requirements.unit;
 
   // unit required and valid or unit not required
-  if ((requiredUnit && requiredUnit === participantUnit) || !requiredUnit)
-    return ParticipantState.CONFIRMED;
-  return ParticipantState.REJECTED;
+  if (requiredUnit) {
+    const found = participantUnit.some((unit) => unit === requiredUnit);
+    return found ? ParticipantState.CONFIRMED : ParticipantState.REJECTED;
```
Comment (marciammart) [inline packages/node-app/src/utils/validations.utils.ts:69]: No need to loop through all user units
```suggestion
    return participantUnits.includes(requiredUnit)
      ? ParticipantState.CONFIRMED
      : ParticipantState.REJECTED;
```

File: packages/node-app/src/models/user.model.ts [hunk 6]
```diff
@@ -28,7 +28,12 @@ const userSchema = new Schema<User>({
     required: true,
     default: UserRole.USER,
   },
-  unit: { type: String, enum: Object.values(Unit), required: false },
+  unit: {
```
Comment (marciammart) [inline packages/node-app/src/models/user.model.ts:31]:
```suggestion
  units: {
```

File: packages/react-app/src/services/giveawayparticipation.ts [hunk 7]
```diff
@@ -81,7 +81,7 @@ const meetRequirements = async (
   const unit = giveaway.requirements.unit;
   const locationId = giveaway.requirements.location;
 
-  if (unit && userInfo.unit !== unit) {
+  if (unit && !userInfo.unit.some((u) => u === unit)) {
```
Comment (marciammart) [inline packages/react-app/src/services/giveawayparticipation.ts:84]:
```suggestion
  if (unit && !userInfo.unit.includes(unit)) {
```

**Review by marciammart — CHANGES_REQUESTED** [review]

File: packages/node-app/src/controllers/giveaway.controller.ts [hunk 8]
```diff
@@ -146,14 +150,22 @@ export const updateGiveaway = async (req: Request, res: Response) => {
 
 export const addParticipant = async (req: Request, res: Response) => {
   try {
+    // validate user
+    const participantEmail = req.body.id;
+    const user = req.user;
+
+    if (user.role !== UserRole.ADMIN && participantEmail !== user.email) {
```
[the remark this answers is not in the document]
Comment (goncaloadolfo) [reply packages/node-app/src/controllers/giveaway.controller.ts:157]: Admin users can add any user, non-admin users can only add them selfs

File: packages/node-app/src/tests/__utils__/helper.utils.ts [hunk 9]
```diff
@@ -1,10 +1,22 @@
...
-    mockedVerifyToken.mockImplementation(async (req, res, next) => next());
+    mockedVerifyToken.mockImplementation(async (req, res, next) => {
+      const user = await User.create({
+        email: 'name@domain.com',
+        name: 'name user',
+      });
+      req.user = user;
+      next();
+    });
+    mockedVerifyAdmin.mockImplementation(async (req, res) => {
```
[the remark this answers is not in the document]
Comment (goncaloadolfo) [reply packages/node-app/src/tests/__utils__/helper.utils.ts:17]: Yes, we also need to mock this function

File: packages/node-app/src/utils/auth.utils.ts (hunk 4 above)
Comment (goncaloadolfo) [reply packages/node-app/src/utils/auth.utils.ts:22]: There is one, but does not provide the information that we need, such as skills and archived

**Review by marciammart — APPROVED** [review]

**Pull request merged**